packages/hardhat/contracts/periphery/ControllerHelper.sol

```diff
@@ -143,30 +159,94 @@ contract ControllerHelper is FlashControllerHelper, IERC721Receiver {
      * @param _wPowerPerpAmountToBurn amount of WPowerPerp to burn
      * @param _wPowerPerpAmountToBuy amount of WPowerPerp to buy
      * @param _collateralToWithdraw amount of collateral to withdraw from vault
-     * @param _maxToPay max amount of collateral to pay for WPowerPerp token
+     * @param _collateralToBuyWith amount of collateral from vault to use to buy long
      */
     function flashswapWBurnBuyLong(
         uint256 _vaultId,
         uint256 _wPowerPerpAmountToBurn,
         uint256 _wPowerPerpAmountToBuy,
         uint256 _collateralToWithdraw,
-        uint256 _maxToPay
+        uint256 _collateralToBuyWith
     ) external payable {
-        require(_maxToPay <= _collateralToWithdraw.add(msg.value), "Not enough collateral");
+        require(_collateralToBuyWith <= _collateralToWithdraw.add(msg.value), "Not enough collateral");
 
         _exactOutFlashSwap(
             weth,
             wPowerPerp,
             IUniswapV3Pool(wPowerPerpPool).fee(),
             _wPowerPerpAmountToBurn.add(_wPowerPerpAmountToBuy),
-            _maxToPay,
+            _collateralToBuyWith.add(msg.value),
             uint8(FLASH_SOURCE.FLASH_W_BURN),
-            abi.encodePacked(_vaultId, _wPowerPerpAmountToBurn, _wPowerPerpAmountToBuy, _collateralToWithdraw)
+            abi.encodePacked(
+                _vaultId,
+                _wPowerPerpAmountToBurn,
+                _wPowerPerpAmountToBuy,
+                _collateralToWithdraw,
+                _collateralToBuyWith.add(msg.value)
+            )
         );
 
         emit FlashWBurn(msg.sender, _vaultId, _wPowerPerpAmountToBurn, _collateralToWithdraw, _wPowerPerpAmountToBuy);
     }
 
+    /**
+     * @notice mint WPowerPerp and LP into Uniswap v3 pool
+     * @param _vaultId vault ID
+     * @param _wPowerPerpAmount amount of WPowerPerp token to mint
+     * @param _collateralToMint collateral to use for minting
+     * @param _collateralToLP collateral to use for LPing
+     * @param _lowerTick LP lower tick
+     * @param _upperTick LP upper tick
+     */
+    function batchMintLp(
+        uint256 _vaultId,
+        uint256 _wPowerPerpAmount,
+        uint256 _collateralToMint,
+        uint256 _collateralToLP,
+        uint256 _amount0Min,
+        uint256 _amount1Min,
+        uint256 _deadline,
+        int24 _lowerTick,
+        int24 _upperTick
+    ) external payable {
+        require(msg.value == _collateralToMint.add(_collateralToLP), "Wrong ETH sent");
```

Comment:
I think this needs to take into account _amount0Min or _amount1Min for extra ETH amount that might be needed.

When LPing the asset composition may change slightly from when tx sent and mined, so you might need a bit extra ETH or oSQTH (extra oSQTH is a problem, see below).

I also think that we need to ensure that the LP is always created with no more than _wPowerPerpAmount (as we mint a fixed amount), can we do this with the _amount0Min or _amount1Min?

Reply:
The alternative is that we mint extra oSQTH and then burn when unused.

Reply:
Not sure I get ur comment, _amount0Min and _amount1Min used to check for slippage, making sure amounts deposited in LP are greater or equal.

`require(amount0 >= params.amount0Min && amount1 >= params.amount1Min, 'Price slippage check')`